components/server/ee/src/prebuilds/bitbucket-server-service.ts

```diff
@@ -0,0 +1,123 @@
+import { inject, injectable } from "inversify";
+import { BitbucketServerApi } from "../../../src/bitbucket-server/bitbucket-server-api";
+import { AuthProviderParams } from "../../../src/auth/auth-provider";
+import { BitbucketServerContextParser } from "../../../src/bitbucket-server/bitbucket-server-context-parser";
+import { Config } from "../../../src/config";
+import { TokenService } from "../../../src/user/token-service";
+import { BitbucketServerApp } from "./bitbucket-server-app";
+
+@injectable()
+export class BitbucketServerService extends RepositoryService {
+    static PREBUILD_TOKEN_SCOPE = "prebuilds";
+
+    @inject(BitbucketServerApi) protected api: BitbucketServerApi;
+    @inject(Config) protected readonly config: Config;
+    @inject(AuthProviderParams) protected authProviderConfig: AuthProviderParams;
+    @inject(TokenService) protected tokenService: TokenService;
+    @inject(BitbucketServerContextParser) protected contextParser: BitbucketServerContextParser;
+
+    async getRepositoriesForAutomatedPrebuilds(user: User): Promise<ProviderRepository[]> {
+        const repos = await this.api.getRepos(user, { limit: 100, permission: "REPO_ADMIN" });
+        return (repos.values || []).map((r) => {
+            const cloneUrl = r.links.clone.find((u) => u.name === "http")?.href!;
+            // const webUrl = r.links?.self[0]?.href?.replace("/browse", "");
+            const accountAvatarUrl = this.api.getAvatarUrl(r.project.key);
+            return <ProviderRepository>{
+                name: r.name,
+                cloneUrl,
+                account: r.project.key,
+                accountAvatarUrl,
+                // updatedAt: TODO(at): this isn't provided directly
+            };
+        });
+    }
+
+    async canInstallAutomatedPrebuilds(user: User, cloneUrl: string): Promise<boolean> {
+        const { host, repoKind, owner, repoName } = await this.contextParser.parseURL(user, cloneUrl);
+        if (host !== this.authProviderConfig.host) {
+            return false;
+        }
+
+        const identity = user.identities.find((i) => i.authProviderId === this.authProviderConfig.id);
+        if (!identity) {
+            console.error(
+                `Unexpected call of canInstallAutomatedPrebuilds. Not authorized with ${this.authProviderConfig.host}.`,
+            );
+            return false;
+        }
+
+        try {
+            await this.api.getWebhooks(user, { repoKind, repositorySlug: repoName, owner });
+            // reading webhooks to check if admin scope is provided
+        } catch (error) {
+            return false;
+        }
+
+        const permission =
+            (await this.api.getPermission(user, { username: identity.authName, repoKind, owner, repoName })) ||
+            (await this.api.getPermission(user, { username: identity.authName, repoKind, owner }));
+
+        if (permission && ["REPO_ADMIN", "PROJECT_ADMIN"].indexOf(permission) !== -1) {
+            return true;
+        }
```

Comment:
[nit] It can be sometimes helpful to communicate the reasoning for a particular check though a method name. For example:
```
if (hasPermissionToCreateWebhooks(permission) {
```